### PR TITLE
Replaced macros and traits with const generics, made decompose() safe, and added error type

### PR DESCRIPTION
**Pull request by Dicklessgreat**

### CODE REVIEW

File: src/sipo.rs [hunk 1]
```diff
@@ -38,109 +49,103 @@ impl OutputPin for ShiftRegisterPin<'_>
...
-            }
+impl<Pin1, Pin2, Pin3, const N: usize> ShiftRegister<Pin1, Pin2, Pin3, N>
+where
+    Pin1: OutputPin,
+    Pin2: OutputPin,
+    Pin3: OutputPin,
+{
+    /// Creates a new SIPO shift register from clock, latch, and data output pins
+    pub fn new(clock: Pin1, latch: Pin2, data: Pin3) -> Self {
+        ShiftRegister {
```
Comment (burrbull) [inline src/sipo.rs:82]:
```suggestion
        assert!(N > 0 && N % 8 == 0);
        ShiftRegister {
```

Comment (Dicklessgreat) [reply src/sipo.rs:82]: I also think that.
But do all shift registers are 8bit or multiple of 8-bit?
Yes, I also think all are!!:) but not 100% sure.

**Comment by SebastianJL** [comment]
LGTM, but maybe let @hacknus test it first on hardware, if he has time.

**Comment by Dicklessgreat** [comment]
> LGTM, but maybe let @hacknus test it first on hardware, if he has time.

OK.
Sample code works on my machines(Pico and SN74HC595N)

**Comment by SebastianJL** [comment]
Well I guess then it's up to @JoshMcguigan if he wants to wait.

**Comment by JoshMcguigan** [comment]
Thanks for the contribution and reviews!

**Pull request merged**